### E-Commerce US Dataset

\Importing Necessary Libraries

In [4]:
import numpy as np
import pandas as pd
import os
import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 200)

\ Configuration - paths

In [5]:
RAW_PATH = r"c:\NG\E-Commerce US dataset\E-Commerce-US-dataset\data\raw\\"

FILES = [
    "customers.csv", "orders.csv", "order_items.csv",
    "order_payments.csv", "order_reviews.csv",
    "products.csv", "sellers.csv", "geolocation.csv"
]

In [6]:
orders     = pd.read_csv(RAW_PATH + "orders.csv")
customers   = pd.read_csv(RAW_PATH + "customers.csv")
order_items = pd.read_csv(RAW_PATH + "order_items.csv")
order_payments   = pd.read_csv(RAW_PATH + "order_payments.csv")
order_reviews    = pd.read_csv(RAW_PATH + "order_reviews.csv")
products    = pd.read_csv(RAW_PATH + "products.csv")
sellers     = pd.read_csv(RAW_PATH + "sellers.csv")
geolocation     = pd.read_csv(RAW_PATH + "geolocation.csv")

\Data Inspection

In [7]:
def explore_table(name, df):
    print("=" * 60)
    print(f"TABLE: {name}  |  Shape: {df.shape}")
    print("=" * 60)
    print("\n--- First 3 rows ---")
    print(df.head(3))
    print("\n--- Column dtypes ---")
    print(df.dtypes)
    print("\n")


In [8]:
explore_table("orders", orders)

TABLE: orders  |  Shape: (1000000, 8)

--- First 3 rows ---
                               order_id                           customer_id order_status order_purchase_timestamp    order_approved_at order_delivered_carrier_date order_delivered_customer_date  \
0  720fd9e9-cdeb-4dec-a187-f71586eb085a  1e2e2881-a0eb-4cb0-829f-a566e810d05f     canceled      2025-12-27 07:07:20  2025-12-27 08:33:20                          NaN                           NaN   
1  c0142972-63fa-4af2-8070-f583ab769847  380b7418-308c-4bf7-b2bd-3ee446cb9ea6    delivered      2019-06-07 19:30:44  2019-06-08 05:08:44          2019-06-09 05:08:44           2019-06-14 05:08:44   
2  11bdf634-2b87-4d37-8d76-be1e7aff8f3b  89b0d980-868f-478d-a63b-5fea5f265f4f    delivered      2023-04-02 14:39:33  2023-04-02 23:42:33          2023-04-03 23:42:33           2023-04-08 23:42:33   

  order_estimated_delivery_date  
0           2026-01-04 08:33:20  
1           2019-06-16 05:08:44  
2           2023-04-10 23:42:33  

--- Co

In [9]:
explore_table("order_items",order_items)

TABLE: order_items  |  Shape: (2199819, 8)

--- First 3 rows ---
                               order_id  order_item_id                            product_id                             seller_id  shipping_limit_date   price  freight_value  discount_rate
0  720fd9e9-cdeb-4dec-a187-f71586eb085a              1  89804d82-33a1-4558-8fa2-9b0252a2a406  36c2b043-ccf4-4d9c-a0bb-4b8d28737fd7  2025-12-30 07:07:20  916.90          45.47            0.0
1  720fd9e9-cdeb-4dec-a187-f71586eb085a              2  ac10b78c-342e-4b50-9ea9-bddde7db2e79  09936504-415b-4f90-a5b8-ad15a5be67d0  2025-12-30 07:07:20  817.14         172.48            0.0
2  720fd9e9-cdeb-4dec-a187-f71586eb085a              3  f18e6310-0e75-4b7f-bc20-90ac1e7bd466  7d336012-7101-4a66-b6bc-c269620d8df9  2025-12-30 07:07:20   37.04          94.98            0.0

--- Column dtypes ---
order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        s

In [10]:
explore_table("customers",customers)

TABLE: customers  |  Shape: (1000000, 9)

--- First 3 rows ---
                            customer_id                    customer_unique_id  customer_name customer_gender  customer_age  customer_zip_code_prefix customer_city customer_state customer_segment
0  1e2e2881-a0eb-4cb0-829f-a566e810d05f  d45cdff2-5195-41e2-a0e5-6fe597e378dd    James Ramos               M            47                     85037       Phoenix             AZ         Consumer
1  380b7418-308c-4bf7-b2bd-3ee446cb9ea6  35cee471-325e-4ad2-8e4e-7b169dc6df81  Brian Jackson               M            18                     90060   Los Angeles             CA         Consumer
2  89b0d980-868f-478d-a63b-5fea5f265f4f  c820d3a4-edd4-4b32-9605-22d97dcc8e5e    Debra Jones               F            61                     30398       Atlanta             GA         Consumer

--- Column dtypes ---
customer_id                   str
customer_unique_id            str
customer_name                 str
customer_gender               st

In [11]:
explore_table("order_reviews",order_reviews)

TABLE: order_reviews  |  Shape: (933748, 7)

--- First 3 rows ---
                              review_id                              order_id  review_score review_comment_title                          review_comment_message review_creation_date  \
0  0ba49642-47bb-4817-9cc2-bb9ded7de7c5  c0142972-63fa-4af2-8070-f583ab769847             5              Perfect     The quality is amazing. Delivery was quick.  2019-06-19 05:08:44   
1  4e540350-1d67-4f18-bfbb-66e45e16bbc5  11bdf634-2b87-4d37-8d76-be1e7aff8f3b             5                 Good  The service is excellent. Super fast shipping.  2023-04-12 23:42:33   
2  9bf68eb6-676d-4759-956e-564b7b90410c  d29b58ad-af2a-47b4-9214-b2664fb1fdba             5              Perfect           The item is fantastic. Arrived early!  2023-02-08 15:26:24   

  review_answer_timestamp  
0     2019-06-19 07:08:44  
1     2023-04-13 01:42:33  
2     2023-02-08 17:26:24  

--- Column dtypes ---
review_id                    str
order_id                 

In [12]:
explore_table("order_payments",order_payments)

TABLE: order_payments  |  Shape: (1149371, 5)

--- First 3 rows ---
                               order_id  payment_sequential payment_type  payment_installments  payment_value
0  720fd9e9-cdeb-4dec-a187-f71586eb085a                   1       paypal                     1        2084.01
1  c0142972-63fa-4af2-8070-f583ab769847                   1  credit_card                     6        1406.07
2  11bdf634-2b87-4d37-8d76-be1e7aff8f3b                   1      voucher                     1         911.30

--- Column dtypes ---
order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object




In [13]:
explore_table("products",products)

TABLE: products  |  Shape: (2000, 10)

--- First 3 rows ---
                             product_id product_category_name                  product_name product_brand  product_weight_g  product_length_cm  product_height_cm  product_width_cm    cost   price
0  cdc4381f-fa67-41f2-a7d5-07f330097589                 books  When successful medical beat        Vertex               661                 18                 19                 9   22.45   42.42
1  1f0d68f8-18d7-47c6-8c73-ffd89a2a960a                  auto   Pulse Seat Cover Kit (M231)         Pulse              9662                 52                 12                25  124.84  148.16
2  f68621b0-8316-4cc4-9ba6-af209132acd6           electronics          Nova M247 Headphones          Nova              4287                 12                 18                15  704.26  843.43

--- Column dtypes ---
product_id                   str
product_category_name        str
product_name                 str
product_brand                str
p

In [14]:
explore_table("sellers",sellers)

TABLE: sellers  |  Shape: (500, 8)

--- First 3 rows ---
                              seller_id seller_company_name seller_contact_name seller_contact_gender  seller_contact_age  seller_zip_code_prefix seller_city seller_state
0  82787b9a-3c5a-476a-9f1b-a5fc7d98c8cf         Johnson LLC      Holly Mcdaniel                     F                  57                   60633     Chicago           IL
1  75c29e4b-cc77-45c9-a3b6-7aa713f56126           Hicks LLC       Charles Casey                     M                  33                   30339     Atlanta           GA
2  d2bc11e3-2a8f-4d16-9b04-8db5008730de        Briggs-Brown      Daniel Serrano                     M                  33                   98193     Seattle           WA

--- Column dtypes ---
seller_id                   str
seller_company_name         str
seller_contact_name         str
seller_contact_gender       str
seller_contact_age        int64
seller_zip_code_prefix    int64
seller_city                 str
seller_state

In [15]:
explore_table("geolocation",geolocation)

TABLE: geolocation  |  Shape: (11500, 5)

--- First 3 rows ---
   zip_code_prefix  geolocation_lat  geolocation_lng geolocation_city geolocation_state
0            10093        40.756378       -74.032669         New York                NY
1            10075        40.681693       -73.956079         New York                NY
2            10010        40.642143       -73.969978         New York                NY

--- Column dtypes ---
zip_code_prefix        int64
geolocation_lat      float64
geolocation_lng      float64
geolocation_city         str
geolocation_state        str
dtype: object




**Insights**
- The dataset is split across 8 relational tables linked via `order_id` and `customer_id`; orders and customers are the largest at 1M rows each, while sellers (500) and products (2,000) are compact dimension tables.
- All datetime columns across orders, order_items, and reviews are stored as `object` dtype, so explicit type conversion is needed before any time-based analysis.
- order_reviews (933K) falls short of the 1M orders, meaning ~6.6% of orders have no review — a gap that could bias average-score calculations if unaddressed.
- order_payments (1.15M) exceeds order count (1M), confirming some orders use multiple sequential transactions (e.g., partial voucher + credit card split).
- The geolocation table joins on zip code and covers only 10 distinct cities/states in the full file, so geographic enrichment will be limited to listed metros.

\Smart Relational Sampling

In [16]:
SAMPLE_SIZE=10000
print(f"Config set! Sample size = {SAMPLE_SIZE} rows per logical group")
orders = pd.read_csv(RAW_PATH + "orders.csv", nrows=SAMPLE_SIZE)

order_ids = set(orders["order_id"])
customer_ids = set(orders["customer_id"])

print(f"Loaded orders: {orders.shape[0]} rows")

def load_filtered(filename, key_column, valid_ids):
    df = pd.read_csv(RAW_PATH + filename)
    filtered = df[df[key_column].isin(valid_ids)].reset_index(drop=True)
    return filtered

# order_id 
order_items    = load_filtered("order_items.csv",    "order_id", order_ids)
order_payments = load_filtered("order_payments.csv", "order_id", order_ids)
order_reviews  = load_filtered("order_reviews.csv",  "order_id", order_ids)

# customer_id 
customers = load_filtered("customers.csv", "customer_id", customer_ids)

# order_items
product_ids = set(order_items["product_id"])
seller_ids  = set(order_items["seller_id"])

products = load_filtered("products.csv", "product_id", product_ids)
sellers  = load_filtered("sellers.csv",  "seller_id",  seller_ids)

# zip codes 
zips = set(customers["customer_zip_code_prefix"]) | set(sellers["seller_zip_code_prefix"])
geo = pd.read_csv(RAW_PATH + "geolocation.csv")
geolocation = geo[geo["zip_code_prefix"].isin(zips)].reset_index(drop=True)


Config set! Sample size = 10000 rows per logical group
Loaded orders: 10000 rows


In [17]:
tables = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "order_payments": order_payments,
    "order_reviews": order_reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
}

print("===== SAMPLED DATA SUMMARY =====")
for name, df in tables.items():
    print(f"{name:18s} : {df.shape[0]:>6,} rows x {df.shape[1]} cols")

===== SAMPLED DATA SUMMARY =====
customers          : 10,000 rows x 9 cols
orders             : 10,000 rows x 8 cols
order_items        : 21,838 rows x 8 cols
order_payments     : 11,474 rows x 5 cols
order_reviews      :  9,332 rows x 7 cols
products           :  1,984 rows x 10 cols
sellers            :    500 rows x 8 cols
geolocation        : 11,255 rows x 5 cols


**Insights**
- 10K sampled orders produced 21,838 order_items — an average basket size of ~2.18 items per order, consistent with mid-range e-commerce purchasing behaviour.
- 9,332 reviews for 10,000 orders gives a ~93.3% review rate, which is unusually high for organic feedback and may indicate auto-generated or prompted entries.
- 11,474 payment rows for 10,000 orders means roughly 1 in 7 orders used more than one payment method, making `payment_sequential` a necessary field for revenue aggregation.
- All 500 sellers and 1,984 of 2,000 products appear within this 10K-order slice, so the sample is broadly representative of the full catalog.

\Data Dictionary

In [18]:
def make_data_dictionary(df, table_name):
    rows=[]
    for col in df.columns:
        rows.append({
            "table":table_name,
            "column":col,
            "dtype": str(df[col].dtype),
            "non_null_count": df[col].notna().sum(),
            "null_count": df[col].isna().sum(),
            "unique_values":df[col].nunique(),
            "sample_value":df[col].dropna().iloc[0] if df[col].notna().any() else None
        })
    return pd.DataFrame(rows)

all_dicts=[]
for name,df in tables.items():
    all_dicts.append(make_data_dictionary(df, name))

data_dictionary =pd.concat(all_dicts, ignore_index=True)
data_dictionary

,table,column,dtype,non_null_count,null_count,unique_values,sample_value
0,customers,customer_id,str,10000,0,10000,1e2e2881-a0eb-4cb0-829f-a566e810d05f
1,customers,customer_unique_id,str,10000,0,7898,d45cdff2-5195-41e2-a0e5-6fe597e378dd
2,customers,customer_name,str,10000,0,7510,James Ramos
3,customers,customer_gender,str,10000,0,2,M
4,customers,customer_age,int64,10000,0,53,47
5,customers,customer_zip_code_prefix,int64,10000,0,838,85037
6,customers,customer_city,str,10000,0,10,Phoenix
7,customers,customer_state,str,10000,0,10,AZ
8,customers,customer_segment,str,10000,0,3,Consumer
9,orders,order_id,str,10000,0,10000,720fd9e9-cdeb-4dec-a187-f71586eb085a


**Insights**
- `customer_unique_id` shows 7,898 unique values against 10,000 `customer_id` rows, meaning ~21% of customer IDs map to repeat buyers with multiple order accounts.
- Zero null counts across all 240 column-table combinations in the sample suggest either a very clean source or upstream preprocessing — worth cross-checking on the full 1M-row file.
- `customer_age` spans 53 distinct values and `customer_gender` has exactly 2, confirming binary gender encoding; segment is a low-cardinality categorical useful for cohort splitting.
- `discount_rate` defaults to 0.0 for most visible rows in order_items — its actual distribution and share of discounted orders will need a dedicated frequency check.

\Saving the sample dataset

In [19]:
for name,df in tables.items():
    df.to_csv(rf"C:\NG\E-Commerce US dataset\E-Commerce-US-dataset\data\raw_sample\{name}.csv",index= False)

data_dictionary.to_csv(r"C:\NG\E-Commerce US dataset\E-Commerce-US-dataset\data\raw_sample\data_dictionary.csv",index=False)